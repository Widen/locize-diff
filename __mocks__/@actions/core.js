export const mocks = {
  projectId: jest.fn(),
  projectSlug: jest.fn(),
  apiKey: jest.fn(),
}

const defaults = {
  projectId: 'projectId',
  apiKey: 'apiKey',
  leftVersion: 'latest',
  rightVersion: 'production',
  includeDrafts: 'false',
  ignoreDeletedKeys: 'true',
}

export function getInput(key) {
  return mocks[key]?.() ?? defaults[key]
}

export const setFailed = jest.fn()
