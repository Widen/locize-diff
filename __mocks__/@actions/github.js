export const listFilesMock = jest.fn()
export const createCommentMock = jest.fn()
export const updateCommentMock = jest.fn()
export const listCommentsMock = jest.fn()
export const graphqlMock = jest.fn()

const wrapResponse = () => async () => ({
  data: listCommentsMock() ?? [],
})

export const getOctokit = () => ({
  graphql: graphqlMock,
  issues: {
    createComment: createCommentMock,
    listComments: wrapResponse(listCommentsMock),
    updateComment: updateCommentMock,
  },
  pulls: {
    listFiles: wrapResponse(listFilesMock),
  },
})

export const context = {
  payload: {
    pull_request: {},
  },
  issue: {
    number: 123,
  },
}

export const draftMock = jest.fn()

Object.defineProperty(context.payload.pull_request, 'draft', {
  get: draftMock.mockReturnValue(false),
})
