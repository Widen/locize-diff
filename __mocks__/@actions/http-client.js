export const listResourcesMock = jest.fn()
export const fetchResourceMock = jest.fn()
export const putMock = jest.fn()

export class HttpClient {
  async get(url) {
    return {
      readBody: async () => {
        const mock = url.includes('/download/')
          ? listResourcesMock
          : url.includes('/pull/')
          ? fetchResourceMock
          : null

        return JSON.stringify(mock(url))
      },
    }
  }

  put = putMock
}
