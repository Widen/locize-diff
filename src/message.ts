import { getInput } from '@actions/core'
import { ResourceDiff } from './diff'

export const credits =
  '_Generated by [Locize Diff](https://github.com/Widen/locize-diff)._'

function createDiffRow([key, value]: [string, ResourceDiff['diffs'][number]]) {
  return `| \`${key}\` | ${value.left || ''} | ${value.right || ''} |`
}

function createDiffMessage(diff: ResourceDiff) {
  const leftVersion = getInput('leftVersion')
  const rightVersion = getInput('rightVersion')

  const rows = Object.entries(diff.diffs)
    .sort((a, b) => a[0].localeCompare(b[0]))
    .map(createDiffRow)
    .join('\n')

  return `### \`${diff.key}\`

| Key | \`${leftVersion}\` | \`${rightVersion}\` |
| --- | --- | --- |
${rows}`
}

export function createMessage(diffs: ResourceDiff[]) {
  const projectSlug = getInput('projectSlug')
  const link = projectSlug
    ? ` [Click here](https://locize.app/p/${projectSlug}/settings) to go to your project settings to publish your changes.`
    : ''

  return `Heads up! Looks like there are some differences between your two Locize versions.${link}

${diffs.map(createDiffMessage).join('\n\n')}

${credits}`
}
