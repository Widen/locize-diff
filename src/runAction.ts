import { setFailed } from '@actions/core'
import { context } from '@actions/github'
import { runCommand } from './commands'
import { runDiff } from './commands/diff'

export async function runAction() {
  try {
    if (context.eventName === 'issue_comment') {
      await runCommand()
    } else {
      await runDiff()
    }
  } catch (err) {
    setFailed(err.message)
  }
}
