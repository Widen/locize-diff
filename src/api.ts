import { getInput } from '@actions/core'
import { HttpClient } from '@actions/http-client'
import { BearerCredentialHandler } from '@actions/http-client/auth'
import { Dict, ResourceCollection } from './utils/types'

const authHandler = new BearerCredentialHandler(getInput('apiKey'))
const client = new HttpClient(undefined, [authHandler])

export function getJSON<T>(url: string): Promise<T> {
  return client
    .get(url)
    .then((res) => res.readBody())
    .then((res) => JSON.parse(res))
}

interface ListResourcesResponse {
  key: string
  lastModified: string
  size: number
  url: string
}

export async function listResources(
  projectId: string,
  version: string
): Promise<ListResourcesResponse[]> {
  return getJSON(`https://api.locize.app/download/${projectId}/${version}`)
}

export async function collectResources(
  projectId: string,
  version: string
): Promise<ResourceCollection[]> {
  const resources = await listResources(projectId, version)
  const promises = resources.map(async (resource) => {
    const [language, namespace] = resource.url.split('/').slice(-2)
    const url = resource.url.replace(projectId, `pull/${projectId}`)

    return {
      key: `${language}/${namespace}`,
      resources: await getJSON<Dict>(url),
    }
  })

  return Promise.all(promises)
}

export async function updateTranslations(
  key: string,
  updates: Record<string, string | null>
) {
  const projectId = getInput('projectId')
  const version = getInput('rightVersion')

  console.log(
    'Updating',
    `https://api.locize.app/update/${projectId}/${version}/${key}`,
    updates
  )

  await client.post(
    `https://api.locize.app/update/${projectId}/${version}/${key}`,
    JSON.stringify(updates)
  )
}
