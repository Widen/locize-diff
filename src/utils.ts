import { getJSON, listResources } from './api'
import { Dict, ResourceCollection } from './types'

export async function collectResources(
  projectId: string,
  version: string
): Promise<ResourceCollection[]> {
  const resources = await listResources(projectId, version)
  const promises = resources.map(async (resource) => {
    const [language, namespace] = resource.url.split('/').slice(-2)

    return {
      key: `${language}/${namespace}`,
      resources: await getJSON<Dict>(resource.url),
    }
  })

  return Promise.all(promises)
}
