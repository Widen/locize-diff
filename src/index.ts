import { getInput, setFailed } from '@actions/core'
import { context, getOctokit } from '@actions/github'
import { getComment, minimizeComment } from './comments'
import { diffResources } from './diff'
import { createMessage } from './message'
import { collectResources } from './utils'

async function main() {
  const projectId = getInput('projectId')
  const leftVersion = getInput('leftVersion')
  const rightVersion = getInput('rightVersion')
  const includeDrafts = getInput('includeDrafts') !== 'true'

  if (includeDrafts && context.payload.pull_request?.draft) {
    return
  }

  try {
    const octokit = getOctokit(getInput('token'))
    const left = await collectResources(projectId, leftVersion)
    const right = await collectResources(projectId, rightVersion)
    const diffs = diffResources(left, right)
    const comment = await getComment()

    if (diffs.length) {
      const req = {
        ...context.repo,
        body: createMessage(diffs),
        issue_number: context.issue.number,
      }

      if (comment) {
        if (comment.body !== req.body) {
          await octokit.issues.updateComment({ ...req, comment_id: comment.id })
        }
      } else {
        await octokit.issues.createComment(req)
      }
    }

    // If the comment exists and there are no longer any diffs, we minimize the
    // comment so it no longer shows in the GitHub UI.
    if (comment && !diffs.length) {
      await minimizeComment(comment.node_id)
    }
  } catch (err) {
    setFailed(err.message)
  }
}

main()
