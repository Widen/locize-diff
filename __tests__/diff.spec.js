jest.mock('@actions/core')
jest.mock('@actions/github')
jest.mock('@actions/http-client')

import { mocks } from '@actions/core'
import {
  createCommentMock,
  listCommentsMock,
  updateCommentMock,
  graphqlMock,
} from '@actions/github'
import { createMessage } from '../src/message'
import { runAction } from '../src/runAction'
import { mockListResources, mockFetchResource } from './utils'

beforeEach(() => {
  jest.resetAllMocks()
})

it('should add a comment when there are diffs', async () => {
  mocks.projectId.mockReturnValue('project-a')
  mockListResources((version) => [`project-a/${version}/en-US/translation`])
  mockFetchResource(
    { 'en-US/translation': { foo: 'bar', hi: 'ho' } },
    { 'en-US/translation': { foo: 'baz', hi: 'ho' } }
  )

  await runAction()
  expect(createCommentMock).toHaveBeenCalledTimes(1)
  expect(createCommentMock.mock.calls[0][0].issue_number).toBe(123)
  expect(createCommentMock.mock.calls[0][0].body).toMatchSnapshot()
})

it('should not a comment when there are diffs', async () => {
  mockListResources((version) => [`projectId/${version}/en-US/translation`])
  mockFetchResource(
    { 'en-US/translation': { foo: 'bar', hi: 'ho' } },
    { 'en-US/translation': { foo: 'bar', hi: 'ho' } }
  )

  await runAction()
  expect(createCommentMock).not.toHaveBeenCalled()
})

const sampleComment = createMessage([
  {
    diffs: {
      foo: { left: 'bar', right: 'baz' },
    },
    key: 'en-US/translation',
  },
])

it('should not comment if the comment already exists', async () => {
  listCommentsMock.mockReturnValue([{ body: sampleComment }])
  mockListResources((version) => [`projectId/${version}/en-US/translation`])
  mockFetchResource(
    { 'en-US/translation': { foo: 'bar', hi: 'ho' } },
    { 'en-US/translation': { foo: 'baz', hi: 'ho' } }
  )

  await runAction()
  expect(createCommentMock).not.toHaveBeenCalled()
  expect(updateCommentMock).not.toHaveBeenCalled()
})

it('should update the existing comment if it exists', async () => {
  listCommentsMock.mockReturnValue([{ body: sampleComment, id: 2 }])
  mockListResources((version) => [`projectId/${version}/en-US/translation`])
  mockFetchResource(
    { 'en-US/translation': { foo: 'bar', hi: 'ho' } },
    { 'en-US/translation': { foo: 'baz', hi: 'howdy' } }
  )

  await runAction()
  expect(createCommentMock).not.toHaveBeenCalled()
  expect(updateCommentMock).toHaveBeenCalledTimes(1)
  expect(updateCommentMock.mock.calls[0][0].comment_id).toBe(2)
  expect(updateCommentMock.mock.calls[0][0].issue_number).toBe(123)
  expect(updateCommentMock.mock.calls[0][0].body).toMatchSnapshot()
})

it('should display diffs for each language/namespace', async () => {
  mockListResources((version) => [
    `projectId/${version}/en-US/common`,
    `projectId/${version}/en-US/translation`,
    `projectId/${version}/zh-CN/ns`,
  ])

  mockFetchResource(
    {
      'en-US/translation': { foo: 'bar', hi: 'ho' },
      'en-US/common': { en: 'common' },
      'zh-CN/ns': {
        a: 'abc',
        b: '123',
        c: 'efg',
      },
    },
    {
      'en-US/translation': { foo: 'baz', hi: 'howdy' },
      'en-US/common': { en: 'not common' },
      'zh-CN/ns': {
        a: '123',
        b: 'abc',
        c: 'efg',
      },
    }
  )

  await runAction()
  expect(createCommentMock).toHaveBeenCalledTimes(1)
  expect(createCommentMock.mock.calls[0][0].issue_number).toBe(123)
  expect(createCommentMock.mock.calls[0][0].body).toMatchSnapshot()
})

it('should minimize the existing comment if there are no longer diffs', async () => {
  listCommentsMock.mockReturnValue([{ body: sampleComment, node_id: 2 }])
  mockListResources((version) => [`projectId/${version}/en-US/translation`])
  mockFetchResource(
    { 'en-US/translation': { foo: 'bar' } },
    { 'en-US/translation': { foo: 'bar' } }
  )

  await runAction()
  expect(graphqlMock).toHaveBeenCalledWith(expect.anything(), { subjectId: 2 })
})

it('should display a link to the product settings if a projectSlug is provided', async () => {
  mocks.projectSlug.mockReturnValue('42od1i754u')
  mockListResources((version) => [`projectId/${version}/en-US/translation`])
  mockFetchResource(
    { 'en-US/translation': { foo: 'bar' } },
    { 'en-US/translation': { foo: 'baz' } }
  )

  await runAction()
  expect(createCommentMock).toHaveBeenCalledTimes(1)
  expect(createCommentMock.mock.calls[0][0].issue_number).toBe(123)
  expect(createCommentMock.mock.calls[0][0].body).toMatchSnapshot()
})
