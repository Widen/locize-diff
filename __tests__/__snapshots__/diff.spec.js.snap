// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add a comment when there are diffs 1`] = `
"Heads up! Looks like there are some differences between your two Locize versions.

### \`en-US/translation\`

| Key | \`latest\` | \`production\` |
| --- | --- | --- |
| \`foo\` | bar | baz |

_Generated by [Locize Diff](https://github.com/Widen/locize-diff)._"
`;

exports[`should update the existing comment if it exists 1`] = `
"Heads up! Looks like there are some differences between your two Locize versions.

### \`en-US/translation\`

| Key | \`latest\` | \`production\` |
| --- | --- | --- |
| \`foo\` | bar | baz |
| \`hi\` | ho | howdy |

_Generated by [Locize Diff](https://github.com/Widen/locize-diff)._"
`;
